runtime/src/fees.rs

```diff
@@ -0,0 +1,242 @@
+///! give back fees
+use frame_support::{
+	traits::{Currency, ExistenceRequirement, Imbalance, IsType, OnUnbalanced, WithdrawReasons},
+	unsigned::TransactionValidityError,
+};
+use frame_system::Config as SystemConfig;
+use pallet_creditcoin::Config as CreditcoinConfig;
+use pallet_transaction_payment::Config as TxPaymentConfig;
+use pallet_transaction_payment::OnChargeTransaction;
+use sp_runtime::{
+	traits::{BlockNumberProvider, Saturating, UniqueSaturatedFrom, Zero},
+	transaction_validity::InvalidTransaction,
+};
+use sp_std::marker::PhantomData;
+
+const BLOCK_HOUR: u32 = 60;
+const BLOCK_YEAR: u32 = BLOCK_HOUR * 24 * 365 + 6 * BLOCK_HOUR;
+
+pub struct CurrencyFeeRedemptionAdapter<C, OU>(PhantomData<(C, OU)>);
+
+impl<C, OU> CurrencyFeeRedemptionAdapter<C, OU> {
+	fn bucketed_year_offset<T: SystemConfig>(at: &T::BlockNumber) -> T::BlockNumber {
+		let base = at.saturating_add(BLOCK_YEAR.into());
+		base - base % UniqueSaturatedFrom::unique_saturated_from(10u32)
+	}
+}
+
+impl<T, C, OU> OnChargeTransaction<T> for CurrencyFeeRedemptionAdapter<C, OU>
+where
+	T: CreditcoinConfig + TxPaymentConfig,
+	C: Currency<<T as SystemConfig>::AccountId>,
+	C::Balance: IsType<T::Balance>,
+	C::PositiveImbalance: Imbalance<C::Balance, Opposite = C::NegativeImbalance>,
+	C::NegativeImbalance: Imbalance<C::Balance, Opposite = C::PositiveImbalance>,
+	OU: OnUnbalanced<C::NegativeImbalance>,
+{
+	type Balance = C::Balance;
+	type LiquidityInfo = Option<C::NegativeImbalance>;
+
+	/// Note: The `fee` already includes the `tip`.
+	fn withdraw_fee(
+		who: &<T>::AccountId,
+		_call: &<T as SystemConfig>::Call,
+		_dispatch_info: &sp_runtime::traits::DispatchInfoOf<<T as SystemConfig>::Call>,
+		fee: Self::Balance,
+		tip: Self::Balance,
+	) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+		if fee.is_zero() {
+			return Ok(None);
+		}
+
+		let withdraw_reason = if tip.is_zero() {
+			WithdrawReasons::TRANSACTION_PAYMENT
+		} else {
+			WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+		};
+
+		match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+			Ok(imbalance) => Ok(Some(imbalance)),
+			Err(_) => Err(InvalidTransaction::Payment.into()),
+		}
```

Comment:
```suggestion
<CurrencyAdapter<Balances, ()> as OnChargeTransaction>::withdraw_fee(who, _call, _dispatch_info, fee, tip)
```

Instead of copying and pasting the currency adapter code, you can just delegate to that implementation.